crates/red_knot_python_semantic/resources/mdtest/unreachable.md

```diff
@@ -0,0 +1,259 @@
+```
+
+```py
+import sys
+
+if sys.version_info >= (3, 11):
+    # TODO: we should not emit an error here
+    # error: [unresolved-import]
+    from typing import Self
+```
+
+#### Checking with Python version 3.12
+
+```toml
+[environment]
+python-version = "3.12"
+```
+
+```py
+import sys
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+```
+
+### `sys.platform` branches
+
+The problem is even more pronounced with `sys.platform` branches, since we don't necessarily have
+the platform information available.
+
+#### Checking with platform `win32`
+
+```toml
+[environment]
+python-platform = "win32"
+```
+
+```py
+import sys
+
+if sys.platform == "win32":
+    sys.getwindowsversion()
+```
+
+#### Checking with platform `linux`
+
+```toml
+[environment]
+python-platform = "linux"
+```
+
+```py
+import sys
+
+if sys.platform == "win32":
+    # TODO: we should not emit an error here
+    # error: [unresolved-attribute]
+    sys.getwindowsversion()
+```
+
+#### Checking without a specified platform
+
+```toml
+[environment]
+# python-platform not specified
+```
+
+```py
+import sys
+
+if sys.platform == "win32":
+    # TODO: we should not emit an error here
+    # error: [possibly-unbound-attribute]
+    sys.getwindowsversion()
+```
+
+#### Checking with platform set to `all`
+
+```toml
+[environment]
+python-platform = "all"
+```
+
+```py
+import sys
+
+if sys.platform == "win32":
+    # TODO: we should not emit an error here
+    # error: [possibly-unbound-attribute]
+    sys.getwindowsversion()
+```
+
+## No false positive diagnostics in unreachable code
+
+In this section, we make sure that we do not emit false positive diagnostics in unreachable code.
+
+### Use of variables in unreachable code
+
+We should not emit any diagnostics for uses of symbols in unreachable code:
+
+```py
+def f():
+    x = 1
+    return
+
+    print("unreachable")
+
+    # TODO: we should not emit an error here; we currently do, since there is no control flow path from this
+    # use of 'x' to any definition of 'x'.
+    # error: [unresolved-reference]
+    print(x)
+```
+
+### Use of variable in nested function
+
+In the example below, since we use `x` in the `inner` function, we use the "public" type of `x`,
+which currently refers to the end-of-scope type of `x`. Since the end of the `outer` scope is
+unreachable, we treat `x` as if it was not defined. This behavior can certainly be improved.
+
+```py
+def outer():
+    x = 1
+
+    def inner():
+        # TODO: we should not emit an error here
+        # error: [unresolved-reference]
+        return x  # Name `x` used when not defined
+    while True:
+        pass
+```
+
+## No diagnostics in unreachable code
+
+In general, no diagnostics should be emitted in unreachable code. The reasoning is that any issues
+inside the unreachable section would not cause problems at runtime. And type checking the
+unreachable code under the assumption that it *is* reachable might lead to false positives:
```

Comment:
I like that reasoning. I assume that only means we won't emit diagnostics. Would we still infer types for those sections? 

It will be interesting to see how we integrate this into the linter... but that's a problem for another day

Reply:
I think we _can_ still infer types for unreachable sections. Whether it's useful is less clear; you are likely to see a lot of `Never` types, since no binding from outside the unreachable code section can reach the unreachable code section.